src/lib/output/SarifOutputFormatter.ts

```diff
@@ -1,18 +1,56 @@
-import { deepCopy } from '../util/Utils';
-import { ENGINE, PMD_VERSION, SFGE_VERSION } from '../../Constants';
-import {Rule, RuleResult, RuleViolation} from '../../types';
-import * as url from 'url';
-import { RuleCatalog } from '../services/RuleCatalog';
-import { Controller } from '../../Controller';
-import {isPathlessViolation} from '../util/Utils';
-import { ESLint } from 'eslint';
+import {OutputFormatter, Results} from "./Results";
+import {FormattedOutput, Rule, RuleResult, RuleViolation} from "../../types";
+import {Location, Log, Notification, Region, ReportingDescriptor, Result, Run} from "sarif";
+import {deepCopy, isPathlessViolation} from "../util/Utils";
+import * as url from "url";
+import {RuleCatalog} from "../services/RuleCatalog";
+import {ESLint} from "eslint";
+import {ENGINE, PMD_VERSION, SFGE_VERSION} from "../../Constants";
+import {Controller} from "../../Controller";
 import * as retire from 'retire';
-import {Location, Log, Notification, Region, Result, Run, ReportingDescriptor} from 'sarif';
 
 const ERROR = 'error';
 const WARNING = 'warning';
 const SINK_VERTEX_MESSAGE = "Sink vertex found here.";
 
+export class SarifOutputFormatter implements OutputFormatter {
+
+	/**
+	 * Convert an array of RuleResults to a sarif document. The rules are separated by engine name.
+	 * A new "run" object is created for each engine that was run
+	 */
+	public async format(results: Results): Promise<FormattedOutput> {
+		// Obtain the catalog and pass it in, this avoids multiple initializations
+		// when waiting for promises in parallel
+		const catalog: RuleCatalog = await Controller.getCatalog();
+		const sarif: Log = {
+			version: '2.1.0',
+			$schema: 'http://json.schemastore.org/sarif-2.1.0',
+			runs: [
+			]
+		};
+
+		// Partition the RuleResults by the engine that generated them. Certain engines may
+		// have multiple RuleResults depending on how the targets were provided to the run command.
+		// Some engines may have no results, these engines should still generate a "run" node
+		const filteredResults: Map<string, RuleResult[]> = new Map<string, RuleResult[]>();
+		for (const engine of results.getExecutedEngines()) {
+			filteredResults.set(engine, []);
+		}
+		for (const r of results.getRuleResults()) {
+			filteredResults.get(r.engine).push(r);
+		}
+
+		// Create a new run object for each engine/results pair
+		for (const [engine, ruleResults] of filteredResults.entries()) {
+			const formatter: SarifFormatter = getSarifFormatter(engine);
```

Comment:
For the sake of keeping the refactor clean, this is fine, but it's a little odd to me that `SarifOutputFormatter` delegates to `SarifFormatter`. Maybe we could combine those classes in the future, or something?

Reply:
Yes I agree. I again just moved stuff around and did minimum code manipulation here. Next time someone is working on that format, I would recommend they restructure it. Just didn't want to add any more noise to these refactorings in this round.